<a href="https://colab.research.google.com/github/manfredialdo/iad-tpi01/blob/main/Manfredi_Aldo_10_Entrega2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## . Datos de identificación

- **Apellido y nombre:** manfredi aldo dario
- **Comisión:** 10
- **Dataset asignado:** hotel booking TPI grupo J.csv
- **Entrega:** Trabajo practico integrador primera entrega
- **Fecha:** Domingo 6 de septiembre de 2026, 23:59 hs.
- **Repositorio:** https://github.com/manfredialdo/iad-tpi01
- **Archivo:** Manfredi_Aldo_10_Entrega2.ipynb

# 🧩 TPI · Entrega 2 · Limpieza, preparación y transformación de datos (OBLIGATORIA)

En esta entrega se debe aplicar al dataset asignado los procedimientos trabajados durante la Unidad N° 2. El objetivo es diagnosticar la calidad de los datos, justificar decisiones de limpieza y construir las transformaciones necesarias para preparar el dataset para las próximas etapas del análisis.

* **Caso de estudio:** Reservas hoteleras
* **Variable objetivo:** `is_canceled`
* **Lenguaje a elección:** Python o R (debes elegir y utilizar un solo lenguaje).
* **⏱️ Tiempo estimado:** 2 pomodoros de 20 minutos (continuidad de la Semana 4).

---

## 💻 Datos del Estudiante e Identificación

> **Atención:** Completa tus datos personales antes de comenzar con la resolución del notebook.

* **Apellido y Nombre:** manfredi aldo dario
* **Comisión:** 10
* **Número de Entrega:** Entrega 2
* **Dataset Asignado:**  dataset J https://tup.sied.utn.edu.ar/pluginfile.php/32860/course/section/4282/hotel%20booking%20TPI%20grupo%20J.csv
* **Problema de Análisis:**
* **Variable Objetivo:** `is_canceled`

---

## 📝 Requisitos Técnicos de Carga

1. **Notebook independiente:** Debe poder ejecutarse desde el inicio de manera fluida sin depender de archivos de entregas anteriores.
2. **Carga reproducible:** Importación de bibliotecas necesarias y carga directa del dataset original.
3. **Copia de trabajo:** Conservar el dataset original intacto en una variable y crear una copia explícita sobre la cual se aplicarán las modificaciones.

---

## 📋 Contenido Obligatorio de la Entrega

1. **Diagnóstico de calidad:** Revisión de valores faltantes, duplicados, tipos de datos, categorías inconsistentes, valores inválidos y posibles valores atípicos, manteniendo la bitácora correspondiente.
2. **Transformación de variables:** Aplicación justificada de conversiones temporales, codificación, discretización, normalización o estandarización cuando resulte pertinente.
3. **Variables derivadas:** Creación de nuevas columnas útiles para el problema (ej. duración total de la estadía, cantidad total de huéspedes, importe estimado, reserva familiar o segmentos de anticipación).
4. **Bitácora del proceso:** Registro estructurado con las variables transformadas, el problema detectado, la decisión aplicada y su justificación.
5. **Dataset preparado:** Exposición del estado final del conjunto de datos y explicación de cómo las decisiones tomadas lo dejan listo para análisis posteriores.

> 📌 **Criterio metodológico:** No se espera que apliques todas las transformaciones posibles ni que elimines automáticamente cada valor inusual. Cada procedimiento debe responder a un problema o necesidad concreta y estar acompañado por una justificación.

---

## 🔎 Variables Sugeridas para Revisión

| Variable(s) | Aspecto a revisar / Transformación |
| :--- | :--- |
| `children`, `adults`, `babies` | Cantidad total de huéspedes e Inconsistencias. |
| `stays_in_weekend_nights`, `stays_in_week_nights` | Duración total de la estadía. |
| `adr` | Valores faltantes, negativos, ceros o atípicos. |
| `lead_time` | Anticipación de la reserva y valores elevados. |
| `country`, `agent`, `company` | Presencia e interpretación de valores ausentes. |
| `hotel`, `market_segment`, `distribution_channel`, `deposit_type`, `customer_type` | Revisión y posible codificación de variables categóricas. |
| **Variables de fecha de llegada** | Formato de fecha y construcción de variables temporales. |

---

## ✅ Lista de Comprobación (Checklist de Entrega)

- [ ] ¿El notebook identifica al estudiante, la comisión y la entrega?
- [ ] ¿Puede ejecutarse de principio a fin sin depender de otro notebook?
- [ ] ¿Conserva el dataset original intacto y aplica los cambios sobre una copia?
- [ ] ¿Revisa los principales problemas de calidad de los datos?
- [ ] ¿Justifica formalmente cada decisión de limpieza y transformación?
- [ ] ¿Las variables derivadas están correctamente construidas y calculadas?
- [ ] ¿Combina explicaciones en texto con bloques de código y sus resultados?
- [ ] ¿Explica detalladamente cómo quedó preparado el dataset final?

---

## ⚠️ Condiciones y Pautas de Entrega

* **Estado:** Obligatoria
* **Habilitación:** Lunes 31 de agosto de 2026, 00:00 hs.
* **Fecha límite de entrega:** Domingo 6 de septiembre de 2026, 23:59 hs.
* **Formato:** Único archivo `.ipynb`.
* **Tamaño máximo:** 20 MB.
* **Calificación mínima de aprobación:** 6 / 10 puntos.
* **Intentos permitidos:** Hasta 2 intentos (la reaparición será habilitada por el tutor cuando corresponda).
* **Estructura de nombre del archivo:** `Apellido_Nombre_Comision_Entrega2.ipynb`

> 💡 **Recomendación:** Conserva también una versión editable acumulativa (notebook maestro) donde integres progresivamente las distintas etapas del TPI para la entrega final.

carga del dataset


## 1. Carga Reproducible y Copia de Trabajo
Importamos las librerías necesarias, cargamos el conjunto de datos desde el repositorio oficial y creamos una copia explícita de trabajo para preservar el dataset original intacto... y la funcion paa la bitacora de registro ..

In [1]:
import pandas as pd
import numpy as np

# Carga del dataset original con la URL correcta del repositorio
url = "https://raw.githubusercontent.com/manfredialdo/iad-tpi01/refs/heads/main/hotel%20booking%20TPI%20grupo%20J.csv"
df = pd.read_csv(url)

# Copia explícita de trabajo
df_limpio = df.copy()

# Inicialización de la Bitácora
registros_bitacora = []

def registrar(problema, variable, decision, justificacion):
    registros_bitacora.append({
        "problema_detectado": problema,
        "variable_afectada": variable,
        "decision_tomada": decision,
        "justificacion": justificacion
    })

print(f"Dataset original cargado correctamente: {df_limpio.shape[0]} filas y {df_limpio.shape[1]} columnas.")

Dataset original cargado correctamente: 25000 filas y 32 columnas.


## 2. Diagnóstico Exploratorio Inicial (EDA)
Revisamos las dimensiones generales, tipos de datos y resúmenes estadísticos para entender la estructura de la base antes de aplicar cualquier modificación

In [2]:
print(df_limpio.shape)

(25000, 32)


In [3]:
# A) Reservas sin huéspedes (adults + children + babies == 0)
# A) Identificación de duplicados por combinación de negocio (sin considerar booking_id)
subset_identificacion = [
    'hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'lead_time', 'adults', 'children', 'babies', 'adr', 'reserved_room_type'
]
duplicados_negocio = df_limpio.duplicated(subset=subset_identificacion, keep=False).sum()

registrar(
    problema=f"Se detectaron {duplicados_negocio} registros duplicados por combinación clave de reserva.",
    variable="Atributos combinados de reserva",
    decision="Conservar registros registrando el hallazgo.",
    justificacion="Pueden responder a reservas simultáneas de grupos gestionadas por la misma agencia."
)

# B) Conteo preliminar de reservas sin huéspedes
sin_huespedes = df_limpio[(df_limpio['adults'] + df_limpio['children'].fillna(0) + df_limpio['babies']) == 0]
cant_sin_huespedes = len(sin_huespedes)

if cant_sin_huespedes > 0:
    df_limpio = df_limpio.drop(sin_huespedes.index)
    registrar(
        problema=f"Existencia de {cant_sin_huespedes} reservas sin ningún huésped registrado.",
        variable="adults, children, babies",
        decision="Eliminar filas afectadas del dataset de trabajo.",
        justificacion="Una reserva requiere al menos 1 huésped para tener validez comercial."
    )

# C) Detección de ADR inconsistente (<= 0)
adr_invalidos = df_limpio[df_limpio['adr'] <= 0]
registrar(
    problema=f"Presencia de {len(adr_invalidos)} registros con ADR menor o igual a 0.",
    variable="adr",
    decision="Identificar valores no operativos para aislar en análisis tarifarios.",
    justificacion="Un valor de tarifa <= 0 distorsiona el cálculo del precio medio real."
)

# D) Outliers en adultos
adults_outliers = df_limpio[df_limpio['adults'] > 10]
if len(adults_outliers) > 0:
    registrar(
        problema=f"Valores atípicos en adultos (máximo de {df_limpio['adults'].max()}).",
        variable="adults",
        decision="Registrar inconsistencia de carga.",
        justificacion="Capacidad excede la ocupación estándar de una habitación individual."
    )

Posee 25000 filas y 32 columnas

In [4]:
# Estructura de tipos de datos y conteo de nulos inicial
df_limpio.info()

# Resumen estadístico numérico
display(df_limpio.describe())

# Resumen estadístico categórico
display(df_limpio.describe(include=['object', 'category']))

<class 'pandas.core.frame.DataFrame'>
Index: 24963 entries, 0 to 24999
Data columns (total 32 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   booking_id                      24963 non-null  object 
 1   hotel                           24963 non-null  object 
 2   is_canceled                     24963 non-null  int64  
 3   lead_time                       24963 non-null  int64  
 4   arrival_date_year               24963 non-null  int64  
 5   arrival_date_month              24963 non-null  object 
 6   arrival_date_week_number        24963 non-null  int64  
 7   arrival_date_day_of_month       24963 non-null  int64  
 8   arrival_date                    24963 non-null  object 
 9   stays_in_weekend_nights         24963 non-null  int64  
 10  stays_in_week_nights            24963 non-null  int64  
 11  adults                          24963 non-null  int64  
 12  children                        24963

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,24963.000000,21494.000000,1437.000000,24963.000000,24963.000000,24963.000000,24963.000000
mean,0.370869,104.290870,2024.156552,26.689060,15.806193,0.919641,2.497416,1.857189,0.103233,0.008052,0.032889,0.090414,0.143212,0.226776,86.458128,187.304802,2.273885,101.814708,0.063013,0.567961
std,0.483047,107.102186,0.707480,13.406402,8.747869,0.996042,1.896072,0.567313,0.393490,0.091586,0.178349,0.874116,1.614336,0.636510,111.071408,128.623369,17.075489,47.859785,0.243815,0.791790
min,0.000000,0.000000,2023.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,9.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,18.000000,2024.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.360000,0.000000,0.000000
50%,0.000000,69.000000,2024.000000,27.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,186.000000,0.000000,94.500000,0.000000,0.000000
75%,1.000000,160.000000,2025.000000,37.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,268.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,629.000000,2025.000000,52.000000,31.000000,18.000000,42.000000,40.000000,3.000000,2.000000,1.000000,26.000000,72.000000,18.000000,535.000000,541.000000,391.000000,387.000000,3.000000,5.000000


,booking_id,hotel,arrival_date_month,arrival_date,meal,country,market_segment,distribution_channel,reserved_room_type,assigned_room_type,deposit_type,customer_type
count,24963,24963,24963,24963,24963,24846,24963,24963,24963,24963,24963,24963
unique,24963,2,12,793,5,134,7,4,9,11,3,4
top,HB-045629,City Hotel,August,2023-12-05,BB,PRT,Online TA,TA/TO,A,A,No Deposit,Transient
freq,1,16580,2901,94,19278,10137,11670,20357,18015,15537,21843,18767


 Conclusiones parciales del Análisis Exploratorio de Datos

---

Comportamiento de Cancelación: La variable objetivo is_canceled muestra una tasa de cancelación del 37.0% (mean = 0.3703). La anticipación promedio de reserva (lead_time) es muy elevada, con 104.2 días de media y una mediana de 69 días (llegando a máximos extremos de 629 días). Asimismo, predomina la política No Deposit (87.5% del total con 21,880 registros), facilitando reservas tempranas sin compromiso financiero que derivan en un alto volumen de cancelaciones.

Perfil de Comercialización y Canales: City Hotel concentra la mayor parte de la demanda con 66.5% de los registros (16,614 de 25,000). El canal de distribución dominante es TA/TO (81.5% con 20,384 reservas), impulsado por el segmento Online TA (11,687 registros). El país de origen predominante es Portugal (PRT), representando el 40.8% de los huéspedes con país registrado (10,159 de 24,883).

Composición de Huéspedes y Estancias: El 75.2% de los clientes son de tipo individual (Transient, 18,796 registros). Las reservas promedian 1.85 adultos, mientras que la presencia de niños (0.10) y bebés (0.008) es marginal. Las estancias duran en promedio 3.4 noches totales (2.5 noches de semana y 0.9 de fin de semana). Solo el 3.3% corresponde a huéspedes recurrentes (is_repeated_guest = 0.033).

Métricas Tarifarias y Operativas: La Tarifa Media Diaria (adr) promedia $101.68, situándose el 50% central entre $69.02 (Q1) y $126.00 (Q3). El régimen alimenticio preferido es BB (Bed & Breakfast) con 19,296 menciones. La habitación tipo A es la más demandada (18,038) y asignada (15,545), observándose discrepancias entre ambas columnas que sugieren reasignaciones o upgrades.

Diagnóstico de Calidad de Datos:

Nulos: La columna country registra 117 nulos (0.47%), mientras que agent (3,479 nulos / 13.92%) y company (23,560 nulos / 94.24%) reflejan transacciones sin intermediarios o de clientes particulares.

Valores Atípicos (Outliers): Se detectan picos inusuales como un máximo de 40 adultos en una reserva, estancias de hasta 42 noches de semana, un adr con mínimo de $0.00 y hasta 391 días en lista de espera (days_in_waiting_list).

In [5]:
df_limpio.duplicated().sum()

np.int64(0)

observamos que no existen valores duplicados. Por lo tanto, no necesitamos realizar ninguna acción en este paso y contiuamos con el resto de la limpieza de datos.

In [6]:
df_limpio.isna().sum()

,0
booking_id,0
hotel,0
is_canceled,0
lead_time,0
arrival_date_year,0
arrival_date_month,0
arrival_date_week_number,0
arrival_date_day_of_month,0
arrival_date,0
stays_in_weekend_nights,0


🧩 Análisis de Valores Nulos (`isna().sum()`)

El diagnóstico de datos faltantes muestra un conjunto de datos con **excelente nivel de integridad**, donde **30 de las 33 columnas no presentan ningún valor nulo**.

* **`country` (117 nulos - 0.47%):** Representa una cantidad residual de reservas sin país registrado.
* **`agent` (3,479 nulos - 13.92%):** Indica reservas realizadas directamente por el cliente sin intermediación de una agencia.
* **`company` (23,560 nulos - 94.24%):** Refleja que la gran mayoría de los huéspedes son viajeros particulares y no corporativos.

> **Estrategia de Limpieza:** Se recomienda imputar `agent` y `company` con el valor `0` (indicando ausencia de agente/empresa) e imputar `country` con la moda (`PRT`) o la categoría `'Unknown'`.

## 3. Diagnóstico de Calidad, Justificación y Tratamiento
Siguiendo la metodología exigida por la cátedra:
1. **Diagnóstico:** Se audita cuantitativamente el problema.
2. **Justificación:** Se analiza su impacto operativo e interpretativo en el negocio hotelero.
3. **Tratamiento:** Se ejecuta el código correspondiente y se asienta en la Bitácora.

### A. Revisión de Duplicados Implícitos
* **Diagnóstico:** `df_limpio.duplicated().sum()` devuelve `0` debido a la columna `booking_id` (identificador único)[cite: 5]. Sin embargo, al evaluar la combinación de atributos clave de la reserva (`hotel`, fechas de llegada, `lead_time`, pasajeros, tarifa, tipo de habitación), se identifican **7,332 registros idénticos**[cite: 5].
* **Justificación:** No se deben eliminar automáticamente. En el sector hotelero, las agencias de viajes a menudo cargan lotes de reservas idénticas para diferentes personas de un mismo grupo. Suprimirlas reduciría erróneamente el volumen real de ventas.
* **Tratamiento:** Conservar los registros en el dataset e incluir la observación en la bitácora[cite: 5].

In [7]:
# -------------------------------------------------------------------------
# 1. IDENTIFICACIÓN DE CLAVE ÚNICA Y DUPLICADOS IMPLÍCITOS
# -------------------------------------------------------------------------
# 1. Diagnóstico
subset_identificacion = [
    'hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'lead_time', 'adults', 'children', 'babies', 'adr', 'reserved_room_type'
]
duplicados_negocio = df_limpio.duplicated(subset=subset_identificacion, keep=False).sum()
print(f"Registros duplicados por combinación clave de negocio: {duplicados_negocio}")

# 2. Justificación y Bitácora
registrar(
    problema=f"Se detectaron {duplicados_negocio} registros idénticos por combinación clave de reserva.",
    variable="Atributos combinados de reserva",
    decision="Conservar registros registrando el hallazgo.",
    justificacion="Pueden responder a reservas simultáneas de grupos gestionadas en lote por la misma agencia."
)

Registros duplicados por combinación clave de negocio: 7330


### B. Inconsistencia Logica: Reservas con 0 Huéspedes
* **Diagnóstico:** Evaluamos las reservas donde la suma total de pasajeros (`adults + children + babies`) es igual a `0`[cite: 5].
* **Justificación:** Comercial y operacionalmente es imposible tener una estadía sin ocupantes registrados (*ghost booking* o fallo del sistema). Al no representar una operación real e impedir el cálculo de métricas por pasajero, su eliminación está plenamente justificada.
* **Tratamiento:** Identificar las **37 filas afectadas** y removerlas de `df_limpio`[cite: 5].

In [8]:
# 1. Diagnóstico
sin_huespedes = df_limpio[(df_limpio['adults'] + df_limpio['children'].fillna(0) + df_limpio['babies']) == 0]
cant_sin_huespedes = len(sin_huespedes)
print(f"Cantidad de reservas registradas con 0 huéspedes: {cant_sin_huespedes}")

# 2. Justificación y Bitácora
if cant_sin_huespedes > 0:
    registrar(
        problema=f"Existencia de {cant_sin_huespedes} reservas sin ningún huésped registrado.",
        variable="adults, children, babies",
        decision="Eliminar filas afectadas del dataset de trabajo.",
        justificacion="Una reserva requiere al menos 1 huésped registrado para ser comercialmente válida."
    )
    # 3. Tratamiento
    df_limpio = df_limpio.drop(sin_huespedes.index)

Cantidad de reservas registradas con 0 huéspedes: 0


### C. Tratamiento de Valores Faltantes (Nulos)
* **Diagnóstico:** Se detectan nulos en `agent` (3,479 / 13.92%), `company` (23,560 / 94.24%), `country` (117 / 0.47%) y `children`[cite: 5].
* **Justificación:**
  * `agent` y `company`: Los faltantes señalan que el cliente reservó de manera directa o no corporativa. Imputarlos con `0` mantiene el significado del negocio.
  * `country`: No se imputa con la moda (`PRT`) para no asentar el sesgo de que los datos faltantes sean locales. Se imputa con `'Unknown'`.
  * `children`: Representa ausencia de menores en el registro. Se imputa con `0`.
* **Tratamiento:** Aplicar imputaciones explícitas preservando la estructura del conjunto de datos[cite: 5].

In [9]:
# 1. Diagnóstico de Nulos
print("Conteo inicial de nulos:")
print(df_limpio[['agent', 'company', 'country', 'children']].isna().sum())

# 2 y 3. Tratamiento y Registro en Bitácora

# Agent
df_limpio['agent'] = df_limpio['agent'].fillna(0).astype(int)
registrar("Faltantes en Agente", "agent", "Imputar con 0", "Valor nulo indica venta directa sin intermediación de agencia.")

# Company
df_limpio['company'] = df_limpio['company'].fillna(0).astype(int)
registrar("Faltantes en Empresa", "company", "Imputar con 0", "Valor nulo indica cliente particular no corporativo.")

# Country
df_limpio['country'] = df_limpio['country'].fillna('Unknown')
registrar("Faltantes en País", "country", "Imputar con 'Unknown'", "Evita sesgar la distribución de origen imputando la moda (PRT).")

# Children
df_limpio['children'] = df_limpio['children'].fillna(0).astype(int)
registrar("Faltantes en Niños", "children", "Imputar con 0", "Se asume ausencia de menores ante la falta de registro explícito.")

Conteo inicial de nulos:
agent        3469
company     23526
country       117
children        0
dtype: int64


### D. Revisión de Tarifas e Inconsistencias Extremas
* **Diagnóstico:** Se identifican **415 registros** con tarifa diaria `adr <= 0` y valores extremos como reservas con más de 10 adultos[cite: 5].
* **Justificación:** Las tarifas `<= 0` responden a errores de facturación o promociones bonificadas que distorsionan el precio medio diario ($ADR$). Los picos de huéspedes pueden ser cargas grupales.
* **Tratamiento:** Registrar el hallazgo para aislar o filtrar estas observaciones en análisis de precios, conservando las filas en el dataset operativo para mantener el histórico de ocupación[cite: 5].

In [10]:
# 1. Diagnóstico de Tarifas e Inconsistencias
adr_invalidos = df_limpio[df_limpio['adr'] <= 0]
print(f"Registros con ADR <= 0: {len(adr_invalidos)}")

# 2. Justificación y Bitácora
registrar(
    problema=f"Presencia de {len(adr_invalidos)} registros con ADR menor o igual a 0.",
    variable="adr",
    decision="Identificar valores no operativos para aislar en análisis tarifarios.",
    justificacion="Un valor de tarifa <= 0 distorsiona el cálculo del precio promedio real."
)

# Revisión de outliers en adultos
adults_outliers = df_limpio[df_limpio['adults'] > 10]
if len(adults_outliers) > 0:
    registrar(
        problema=f"Valores atípicos en adultos (máximo detectado: {df_limpio['adults'].max()}).",
        variable="adults",
        decision="Registrar inconsistencia de carga conservando el registro.",
        justificacion="Capacidad excede la ocupación estándar; requiere análisis separado para no distorsionar promedio por habitación."
    )

Registros con ADR <= 0: 415


# 3. IMPUTACIÓN DE VALORES FALTANTES Y CADENAS

In [11]:
df_limpio['agent'] = df_limpio['agent'].fillna(0).astype(int)
registrar("Faltantes en Agente", "agent", "Imputar con 0", "Valor nulo indica venta directa sin agencia.")

df_limpio['company'] = df_limpio['company'].fillna(0).astype(int)
registrar("Faltantes en Empresa", "company", "Imputar con 0", "Valor nulo indica reserva particular no corporativa.")

df_limpio['country'] = df_limpio['country'].fillna('Unknown')
registrar("Faltantes en País", "country", "Imputar con 'Unknown'", "Evita sesgar la distribución imputando la moda.")

df_limpio['children'] = df_limpio['children'].fillna(0).astype(int)
registrar("Faltantes en Niños", "children", "Imputar con 0", "Se asume ausencia de menores ante falta de registro.")

## 4. Transformación de Datos y Creación de Variables Derivadas
Con el objetivo de alimentar los análisis posteriores, construimos las 5 variables derivadas solicitadas por el TPI
1. `arrival_date_parsed`: Conversión formal de la fecha a formato `datetime`

2. `total_guests`: Suma de adultos, niños y bebés

3. `total_stay_nights`: Noches totales (fin de semana + días hábiles)

4. `is_family`: Indicador binario si la reserva incluye menores

5. `estimated_total_cost`: Ingreso total estimado ($ADR \times \text{noches}$)

6. `lead_time_group`: Discretización por tramos temporales de la antelación de reserva

In [12]:
# Crear fecha de llegada unificada en formato datetime
# Parseo correcto usando la columna arrival_date existente (YYYY-MM-DD)
# Conversión formal de fecha
df_limpio['arrival_date_parsed'] = pd.to_datetime(df_limpio['arrival_date'], errors='coerce')

# Creación de variables requeridas
df_limpio['total_guests'] = df_limpio['adults'] + df_limpio['children'] + df_limpio['babies']
df_limpio['total_stay_nights'] = df_limpio['stays_in_weekend_nights'] + df_limpio['stays_in_week_nights']
df_limpio['is_family'] = ((df_limpio['children'] + df_limpio['babies']) > 0).astype(int)
df_limpio['estimated_total_cost'] = df_limpio['total_stay_nights'] * df_limpio['adr']

# Discretización de Lead Time
df_limpio['lead_time_group'] = pd.cut(
    df_limpio['lead_time'],
    bins=[-1, 7, 30, 90, 180, 999],
    labels=['Mismo día/Semana', 'Último mes', '1-3 Meses', '3-6 Meses', 'Más de 6 Meses']
)

print("Variables derivadas creadas y transformaciones completadas.")

Variables derivadas creadas y transformaciones completadas.


In [13]:
# reservas con 0 huespedes?
# Filtrar reservas donde la suma de todos los huéspedes es cero
reservas_sin_huespedes = df_limpio[(df_limpio['adults'] + df_limpio['children'].fillna(0) + df_limpio['babies']) == 0]

print(f"Cantidad de reservas registradas con 0 huéspedes: {len(reservas_sin_huespedes)}")
display(reservas_sin_huespedes[['booking_id', 'hotel', 'is_canceled', 'adults', 'children', 'babies', 'adr']])

Cantidad de reservas registradas con 0 huéspedes: 0


,booking_id,hotel,is_canceled,adults,children,babies,adr


# 5. CREACIÓN DE VARIABLES DERIVADAS

In [14]:
df_limpio['total_guests'] = df_limpio['adults'] + df_limpio['children'] + df_limpio['babies']
df_limpio['total_stay_nights'] = df_limpio['stays_in_weekend_nights'] + df_limpio['stays_in_week_nights']
df_limpio['is_family'] = ((df_limpio['children'] + df_limpio['babies']) > 0).astype(int)
df_limpio['estimated_total_cost'] = df_limpio['total_stay_nights'] * df_limpio['adr']

# Discretización de Lead Time
df_limpio['lead_time_group'] = pd.cut(
    df_limpio['lead_time'],
    bins=[-1, 7, 30, 90, 180, 999],
    labels=['Mismo día/Semana', 'Último mes', '1-3 Meses', '3-6 Meses', 'Más de 6 Meses']
)

In [15]:
df_bitacora = pd.DataFrame(registros_bitacora)
display(df_bitacora)

,problema_detectado,variable_afectada,decision_tomada,justificacion
0,Se detectaron 7332 registros duplicados por co...,Atributos combinados de reserva,Conservar registros registrando el hallazgo.,Pueden responder a reservas simultáneas de gru...
1,Existencia de 37 reservas sin ningún huésped r...,"adults, children, babies",Eliminar filas afectadas del dataset de trabajo.,Una reserva requiere al menos 1 huésped para t...
2,Presencia de 415 registros con ADR menor o igu...,adr,Identificar valores no operativos para aislar ...,Un valor de tarifa <= 0 distorsiona el cálculo...
3,Valores atípicos en adultos (máximo de 40).,adults,Registrar inconsistencia de carga.,Capacidad excede la ocupación estándar de una ...
4,Se detectaron 7330 registros idénticos por com...,Atributos combinados de reserva,Conservar registros registrando el hallazgo.,Pueden responder a reservas simultáneas de gru...
5,Faltantes en Agente,agent,Imputar con 0,Valor nulo indica venta directa sin intermedia...
6,Faltantes en Empresa,company,Imputar con 0,Valor nulo indica cliente particular no corpor...
7,Faltantes en País,country,Imputar con 'Unknown',Evita sesgar la distribución de origen imputan...
8,Faltantes en Niños,children,Imputar con 0,Se asume ausencia de menores ante la falta de ...
9,Presencia de 415 registros con ADR menor o igu...,adr,Identificar valores no operativos para aislar ...,Un valor de tarifa <= 0 distorsiona el cálculo...


In [16]:
# Mostrar la bitácora
df_bitacora = pd.DataFrame(registros_bitacora)
display(df_bitacora)

# Confirmar estado final del dataset
print(f"\nDimensiones finales del dataset preparado: {df_limpio.shape[0]} filas y {df_limpio.shape[1]} columnas.")
display(df_limpio.head())

,problema_detectado,variable_afectada,decision_tomada,justificacion
0,Se detectaron 7332 registros duplicados por co...,Atributos combinados de reserva,Conservar registros registrando el hallazgo.,Pueden responder a reservas simultáneas de gru...
1,Existencia de 37 reservas sin ningún huésped r...,"adults, children, babies",Eliminar filas afectadas del dataset de trabajo.,Una reserva requiere al menos 1 huésped para t...
2,Presencia de 415 registros con ADR menor o igu...,adr,Identificar valores no operativos para aislar ...,Un valor de tarifa <= 0 distorsiona el cálculo...
3,Valores atípicos en adultos (máximo de 40).,adults,Registrar inconsistencia de carga.,Capacidad excede la ocupación estándar de una ...
4,Se detectaron 7330 registros idénticos por com...,Atributos combinados de reserva,Conservar registros registrando el hallazgo.,Pueden responder a reservas simultáneas de gru...
5,Faltantes en Agente,agent,Imputar con 0,Valor nulo indica venta directa sin intermedia...
6,Faltantes en Empresa,company,Imputar con 0,Valor nulo indica cliente particular no corpor...
7,Faltantes en País,country,Imputar con 'Unknown',Evita sesgar la distribución de origen imputan...
8,Faltantes en Niños,children,Imputar con 0,Se asume ausencia de menores ante la falta de ...
9,Presencia de 415 registros con ADR menor o igu...,adr,Identificar valores no operativos para aislar ...,Un valor de tarifa <= 0 distorsiona el cálculo...



Dimensiones finales del dataset preparado: 24963 filas y 38 columnas.


,booking_id,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,arrival_date,stays_in_weekend_nights,...,customer_type,adr,required_car_parking_spaces,total_of_special_requests,arrival_date_parsed,total_guests,total_stay_nights,is_family,estimated_total_cost,lead_time_group
0,HB-035978,Resort Hotel,0,261,2025,April,18,30,2025-04-30,2,...,Contract,42.95,0,1,2025-04-30,2,7,0,300.65,Más de 6 Meses
1,HB-004892,Resort Hotel,1,85,2024,April,14,7,2024-04-07,0,...,Transient-Party,64.00,0,0,2024-04-07,2,3,0,192.00,1-3 Meses
2,HB-076102,City Hotel,1,364,2023,October,42,16,2023-10-16,0,...,Transient-Party,101.50,0,0,2023-10-16,2,2,0,203.00,Más de 6 Meses
3,HB-102752,City Hotel,0,32,2024,December,49,4,2024-12-04,2,...,Transient,114.00,0,0,2024-12-04,1,2,0,228.00,1-3 Meses
4,HB-069831,City Hotel,1,126,2025,June,23,7,2025-06-07,0,...,Transient,89.10,0,1,2025-06-07,2,2,0,178.20,3-6 Meses
